onnxscript/ir/_core.py

```diff
@@ -1977,9 +1977,86 @@ def insert_before(self, node: Node, new_nodes: Iterable[Node] | Node, /) -> None
         self._nodes.insert_before(node, new_nodes)
 
     def sort(self) -> None:
-        """Topologically sort the nodes in the graph."""
-        raise NotImplementedError("Not implemented yet")
-
+        # make sure the graph is a DAG
+        import heapq
+        stk = []
+        nodes = [node for node in self._nodes if isinstance(node, Node)] # screen nodes
+        depth = {node: 0 for node in nodes} # topological depth
+        adjacency_list =  {node: [] for node in nodes} # parent list
+        ind = {node: i for i, node in enumerate(nodes)} # index (priority)
+        for node in nodes:
+            for input_value in node.inputs:
+                p = input_value.producer()
+                if isinstance(p, Node):
+                    adjacency_list[node].append(p)
+                    depth[p] += 1 # parent gain 1 depth with each child
+        pq = [] # priority queue
+        for node in nodes:
+            if depth[node] == 0:
+                heapq.heappush(pq, (ind[node], node))
+        while pq: # use PQ+BFS, archive the leaf with largest index, then check all its parents if they become leaves
+            _, node = heapq.heappop(pq)
+            stk.append(node)
+            for neighbor in adjacency_list[node]:
+                depth[neighbor] -= 1
+                if depth[neighbor] == 0:
+                    heapq.heappush(pq, (ind[neighbor], neighbor))
+        res = _linked_list.DoublyLinkedSet() # reconstruct doubly linked set
```

Comment:
Is the stack a sorted list of nodes? If so I would just call

```python
self.extend(reversed(stk))
```

The linked list implementation ensures uniqueness as a property.

This way we don't need to initialize a new linked list for it.